Exception-Handling.md

```diff
@@ -0,0 +1,209 @@
+# Exception Handling In Rails
+
+## Learning Goals
+At the end of this you should be able to:
+
+-  Explain how exception handling works.
+-  Explain where you would want to use exception handling
+-  Create rescue blocks to "catch" exceptions.
+
+## Errors Errors Errors... How to Handle Them
+
+With any program there are things that can go wrong, a file you are planning to read doesn't exist or is unreadable, the database is unavailable, somehow you try to divide something by zero.  Since there are always things that can go wrong it becomes desirable to have a way to recover without crashing your program or website in front of the user.
+
+Ruby, like many programming languages chooses to handle errors with a technique called *Exception Handling*.  When an error occurs, Ruby generates an object instance called an Exception.  Exceptions are always subclasses of the [Exception](https://ruby-doc.org/core-2.3.1/Exception.html) class. 
+
+[![exceptions](images/exceptions.png)](http://findnerd.com/list/view/Exception-Handling-in-Rails-using-begin-rescue/21677/)
+
+You can mark a section to recover from an error with a rescue block.  If an Exception is not "rescued" it will end the program with an error message.  
+
+Here is an example of an error being handled:
+
+```ruby
+# sample_exception.rb
+begin
+  # Dividing by zero causes an error
+  quotient = 5 / 0
+  puts "Made it past the error"
+rescue
+  quotient = nil
+  puts "Rescued the error and set quotient to nil"
+end
+```
+
+The output would be:
+```bash
+$  ruby sample_exception.rb
+Rescued the error and set quotient to nil
+```
+
+Without the rescue:
+
+```ruby
+# sample_exception.rb
+begin
+  # Dividing by zero causes an error
+  quotient = 5 / 0
+  puts "Made it past the error"
+end
+```
+The output would be:
+```bash
+$  ruby sample_exception.rb
+basic_exception_without_rescue.rb:4:in `/': divided by 0 (ZeroDivisionError)
+	from basic_exception_without_rescue.rb:4:in `<main>'
+```
+
+### Rescuing Specific Types of Exceptions
+
+There are a number of kinds of exceptions both for [regular'ol Ruby](http://blog.nicksieger.com/articles/2006/09/06/rubys-exception-hierarchy/) & [Rails](http://stackoverflow.com/questions/6521544/getting-a-list-of-existing-rails-error-classes-for-re-use-inheritance).  Some types of errors we probably don't want to try to rescue because they are irrecoverable or so unexpected that we shouldn't write code to recover from them.  We can however set up specific blocks to recover from specific types of errors that we can and want to recover from.  
+
+The syntax will look like this:
+
+```ruby
+begin
+  # some code that will create an exception.
+  
+rescue <some kind of exception>
+  # Code to recover from that exception
+
+rescue <some other kind of exception>
+  # Code to recover from another type of exception
+
+else 
+  # other exceptions handled here
```

Comment:
`else` in a `begin ... rescue` statement is actually for code run when no exceptions occurred in the `begin` block. As such, I think it's not really useful to teach (I can't figure out why this would ever be different from putting that code at the end of the `begin` block).

What we should use here instead is an empty `rescue` line, which will handle all other `StandardError` exceptions.